### PR TITLE
(#179) factor out router as independent library

### PR DESCRIPTION
**Pull request by dannypsnl**

### CODE REVIEW

File: handler.go [hunk 1]
```diff
@@ -130,30 +133,31 @@ func (h *handler) fillByCachedUserContexts(contexts []*context.UserContext, reqU
...
-type optionsHandler struct {
-	methods []string
-}
+const ErrorMessageForMethodNotAllowed = "request resource does not support http method '%s'"
 
-func newOptionsHandler() *optionsHandler {
-	return &optionsHandler{
-		methods: make([]string, 0),
-	}
+func notAllowHandler(method string) router.Handler {
```
Comment (shana0440) [inline handler.go]: I think `createNotAllowMethodHandler` is the better name, the current name like a variable, not a function and not sure it does not allow what.

File: filler.go [hunk 2]
```diff
@@ -41,12 +41,12 @@ type (
...
+func newRouteFiller(routes, reqURL []string, routeParams map[int]int, wildcardIndex int) filler {
 	return &routeFiller{
-		routes:           routes,
-		routeParams:      routeParams,
-		reqURL:           reqURL,
-		matchedPathIndex: matchedPathIndex,
+		routes:                    routes,
+		routeParams:               routeParams,
+		reqURL:                    reqURL,
+		wildcardIndexInSplitRoute: wildcardIndex,
```
Comment (shana0440) [inline filler.go]: Great name 🥂

File: handler.go [hunk 3]
```diff
@@ -130,30 +133,31 @@ func (h *handler) fillByCachedUserContexts(contexts []*context.UserContext, reqU
...
 			})
 	}))
 }
+
+func (h *handler) getRoute() string {
+	return h.route
+}
+
+// WildcardIndex implements router.Handler
+func (h *handler) WildcardIndex(i int) error {
```
Comment (shana0440) [inline handler.go]: If this would never return error, I think we should remove the return value.

Comment (dannypsnl) [reply handler.go]: Please take a look at the `router.Handler` interface, I left a comment about why I allow it returns an error, furthermore I'm going to remove this from the interface for the following reasons:
- it depends on the implementation of `router`, `router` split path by `/` for now, but future?
- it's hard to understand the purpose, easy to misuse it

or you can say those issues are the same one: current design make `router` and `handler` depends on another's implementation details.

For now, I won't change this, I think we need another issue to think about the problem.

File: router/router.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+	notAllowHandler       func(method string) Handler
+}
+
+// OptionsHandler interface define an interface that helps router automatically create an OPTION method handler for you
+type OptionsHandler interface {
+	Build(allowMethods string) Handler
+}
+
+type Handler interface {
+	WildcardIndex(int) error // maybe we should fall failed while we emit a wildcard index onto a handler don't do it?
```
Comment (shana0440) [inline router/router.go]: sorry, I cannot understand this comment 🙈  can you explain more?

Comment (dannypsnl) [reply router/router.go]: Ok, I got what the problem is, originally I thought some handler might not want to support Wildcard route if true then they should return an error at here, but seems very wrong concern. Those handlers do not want to support wildcard can just simply empty the implementation or just panic.

Anyway, I would remove that error from `WildcardIndex`

**Pull request merged**